crates/zkvm/zkvm/src/lib.rs

```diff
@@ -204,3 +204,102 @@ impl AggregationInput {
+
+    /// Verifies a Groth16 proof.
+    ///
+    /// # Parameters
+    ///
+    /// * `proof`: A byte slice containing the serialized proof. (TODO: Change `proof: &[u8]` to a
+    ///   `Proof` type.)
+    /// * `verification_key`: A byte slice containing the serialized verification key.
+    /// * `public_params_raw`: A byte slice containing the serialized public parameters.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the proof verifies successfully, or an `anyhow::Error` if verification
+    /// fails.
+    fn verify_groth16(
+        &self,
+        proof: &[u8],
+        verification_key: &[u8],
+        public_params_raw: &[u8],
+    ) -> anyhow::Result<()>;
+
+    /// Reads a Borsh-serialized object from the guest code.
+    ///
+    /// The input is expected to be written with [`ZKVMInputBuilder::write_borsh`].
+    fn read_borsh<T: BorshSerialize + BorshDeserialize>(&self) -> T {
+        let slice = self.read_slice();
+        borsh::from_slice(&slice).expect("borsh serialization failed")
+    }
+
+    /// Commits a Borsh-serializable object to the public values stream.
+    ///
+    /// Values that are committed can be proven as public parameters.
+    fn commit_borsh<T: BorshSerialize + BorshDeserialize>(&self, output: &T) {
+        self.commit_slice(&borsh::to_vec(output).expect("failed borsh serialization"));
+    }
+
+    /// Reads and verifies a committed output from another guest function, deserializing it using
+    /// Serde.
+    ///
+    /// This function is meant to read the committed output of another guest function
+    /// that was written with [`ZkVm::commit`].
+    /// It then verifies the proof against the given verification key digest.
+    ///
+    /// This is equivalent to calling [`ZkVm::read`] and [`ZkVm::verify_proof`], but avoids double
+    /// serialization and deserialization. The function will panic if the proof fails to verify.
+    fn read_verified<T: DeserializeOwned>(&self, vk_digest: &[u32; 8]) -> T {
+        let public_values_raw = self.read_slice();
+        self.verify_proof(vk_digest, &public_values_raw);
+        bincode::deserialize(&public_values_raw).expect("failed bincode deserialization")
+    }
+
+    /// Reads and verifies a committed output from another guest function, deserializing it using
+    /// Borsh.
+    ///
+    /// This function is similar to [`ZkVm::read_verified`], but is intended for guest commitments
+    /// committed via [`ZkVm::commit_borsh`]. The output is expected to be Borsh-serializable.
+    /// It then verifies the proof using the internal verification key context.
+    ///
+    /// This is equivalent to calling [`ZkVm::read_borsh`] and [`ZkVm::verify_proof`], but avoids
+    /// double serialization and deserialization. The function will panic if the proof fails to
+    /// verify.
+    fn read_verified_borsh<T: BorshSerialize + BorshDeserialize>(&self, vk_digest: &[u32; 8]) -> T {
```

Comment:
These two methods feel kinda off to me.  What would be a likely scenario to override these by downstream zkvm impls?  Could they be broken out to an impl fn?

Also see above comment about `_serde`/`_bincode` names.

Reply:
Have updated this to be more consistent.